pkg/reconciler/workspaceindex/server.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workspaceindex
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+	"sort"
+	"strconv"
+
+	"k8s.io/apiserver/pkg/server/healthz"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1/helper"
+)
+
+func (o *Orchestrator) stable() bool {
+	o.activeWorkersLock.RLock()
+	defer o.activeWorkersLock.RUnlock()
+	orgLen, orgWorkers, workspaceLen, workspaceWorkers := o.orgQueue.Len(), o.orgQueueActiveWorkers, o.workspaceQueue.Len(), o.workspaceQueueActiveWorkers
+	klog.Infof("workspaceQueue.Len()=%d, workspaceQueueActiveWorkers=%d, orgQueue.Len()=%d, orgQueueActiveWorkers=%d", workspaceLen, workspaceWorkers, orgLen, orgWorkers)
+	return workspaceWorkers == 0 && orgWorkers == 0 && orgLen == 0 && workspaceLen == 0
```

Comment:
hairy logic. As discussed yesterday: wait for listers to be complete and iterate over the objects once. Much simpler.

Reply:
Can't do that, if we know we have outstanding work ... why bother serving? :)

Reply:
Plus, the logic for "list root once, list all orgs once, list all workspaces in an org once ..." is hard to reason about as well (and totally useless past startup)